File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/BannerPlacement.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import static com.urbanairship.android.layout.model.SafeAreaAware.ignoreSafeAreaFromJson;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.urbanairship.android.layout.model.SafeAreaAware;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

public class BannerPlacement implements SafeAreaAware {
    @NonNull
    private final ConstrainedSize size;
    @Nullable
    private final Margin margin;
    @Nullable
    private final Position position;
    private final boolean ignoreSafeArea;

    @Nullable
    private final Border border;

    @Nullable
    private final Color backgroundColor;

    public BannerPlacement(
        @NonNull ConstrainedSize size,
        @Nullable Margin margin,
        @Nullable Position position,
        boolean ignoreSafeArea,
        @Nullable Border border,
        @Nullable Color backgroundColor
    ) {
        this.size = size;
        this.margin = margin;
        this.position = position;
        this.ignoreSafeArea = ignoreSafeArea;
        this.border = border;
        this.backgroundColor = backgroundColor;
    }

    @NonNull
    public static BannerPlacement fromJson(@NonNull JsonMap json) throws JsonException {
        JsonMap sizeJson = json.opt("size").optMap();
        if (sizeJson.isEmpty()) {
            throw new JsonException("Failed to parse Modal Placement! Field 'size' is required.");
        }
        String positionJson = json.opt("position").optString();
        JsonMap marginJson = json.opt("margin").optMap();
        JsonMap borderJson = json.opt("border").optMap();
        JsonMap backgroundJson = json.opt("background_color").optMap();

        ConstrainedSize size = ConstrainedSize.fromJson(sizeJson);
        Margin margin = marginJson.isEmpty() ? null : Margin.fromJson(marginJson);

        VerticalPosition verticalPosition = VerticalPosition.from(positionJson);
        Position position = new Position(HorizontalPosition.CENTER, verticalPosition);

        boolean ignoreSafeArea = ignoreSafeAreaFromJson(json);

        Border border = borderJson.isEmpty() ? null : Border.fromJson(borderJson);
        Color backgroundColor = backgroundJson.isEmpty() ? null : Color.fromJson(backgroundJson);

        return new BannerPlacement(size, margin, position, ignoreSafeArea, border, backgroundColor);
    }

    @Nullable
    public Margin getMargin() {
        return margin;
    }

    @NonNull
    public ConstrainedSize getSize() {
        return size;
    }

    @Nullable
    public Position getPosition() {
        return position;
    }

    @Override
    public boolean shouldIgnoreSafeArea() {
        return ignoreSafeArea;
    }

    @Nullable
    public Border getBorder() {
        return border;
    }

    @Nullable
    public Color getBackgroundColor() {
        return backgroundColor;
    }

}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/BannerPlacement.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import com.urbanairship.UALog
import com.urbanairship.android.layout.model.SafeAreaAware
import com.urbanairship.android.layout.model.SafeAreaAware.Companion.ignoreSafeAreaFromJson
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalList

public class BannerPlacement public constructor(
    public val size: ConstrainedSize,
    public val margin: Margin?,
    public val position: EdgePosition,
    public val ignoreSafeArea: Boolean,
    public val border: Border?,
    public val backgroundColor: Color?,
    public val transition: BannerTransition?,
    public val swipeToDismiss: Boolean = true,
    public val shadow: Shadow? = null
) : SafeAreaAware {

    override fun shouldIgnoreSafeArea(): Boolean {
        return ignoreSafeArea
    }

    public companion object {

        private const val KEY_SIZE = "size"
        private const val KEY_MARGIN = "margin"
        private const val KEY_POSITION = "position"
        private const val KEY_BORDER = "border"
        private const val KEY_BACKGROUND = "background_color"
        private const val KEY_TRANSITION = "transition"
        private const val KEY_SWIPE_TO_DISMISS = "swipe_to_dismiss"
        private const val KEY_SHADOW = "shadow"
        private const val KEY_SELECTORS = "selectors"

        @JvmStatic
        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): BannerPlacement {
            val content = json.requireMap()

            val shadow = content[KEY_SHADOW]?.let { shadowJson ->
                val resolved = shadowJson.optMap()
                    .optionalList(KEY_SELECTORS)
                    ?.map(ShadowSelector::fromJson)
                    ?.firstOrNull { it.platform == null || it.platform == Platform.ANDROID }
                    ?.shadow
                if (resolved == null) {
                    UALog.w { "Ignoring banner '$KEY_SHADOW'! No usable Android shadow in: $shadowJson" }
                }
                resolved
            }

            val swipeToDismiss = content[KEY_SWIPE_TO_DISMISS]?.let { value ->
                if (value.isBoolean) {
                    value.getBoolean(true)
                } else {
                    UALog.w { "Ignoring banner '$KEY_SWIPE_TO_DISMISS'! Expected a boolean, got: $value" }
                    null
                }
            } ?: true

            return BannerPlacement(
                size = ConstrainedSize.fromJson(content.require(KEY_SIZE)),
                margin = content[KEY_MARGIN]?.let(Margin::fromJson),
                position = EdgePosition.fromJson(content.require(KEY_POSITION)),
                ignoreSafeArea = ignoreSafeAreaFromJson(content),
                border = content[KEY_BORDER]?.let(Border::fromJson),
                backgroundColor = content[KEY_BACKGROUND]?.let(Color::fromJson),
                transition = content[KEY_TRANSITION]?.let(BannerTransition::fromJson),
                swipeToDismiss = swipeToDismiss,
                shadow = shadow
            )
        }
    }
}
